tests/python/pants_test/engine/test_rules.py

```diff
@@ -108,42 +111,49 @@ def test_creation_fails_with_intrinsic_that_overwrites_another_intrinsic(self):
 
 
 class RulesetValidatorTest(unittest.TestCase):
+  def create_validator(self, goal_to_product, intrinsic_entries, root_subject_types, rules):
+    return self.create_native_scheduler(intrinsic_entries, root_subject_types, rules)
+
+  def create_native_scheduler(self, intrinsic_entries, root_subject_types, rules):
+    init_subsystem(Native.Factory)
+    rule_index = RuleIndex.create(rules, intrinsic_entries)
+    native = Native.Factory.global_instance().create()
+    scheduler = WrappedNativeScheduler(native, rule_index, root_subject_types)
+    return scheduler
+
   def test_ruleset_with_missing_product_type(self):
     rules = [(A, (Select(B),), noop)]
-    validator = RulesetValidator(RuleIndex.create(rules, tuple()),
-                                 goal_to_product={},
-                                 root_subject_types={SubA})
+    intrinsic_entries = tuple()
+    root_subject_types = {SubA}
+    scheduler = self.create_native_scheduler(intrinsic_entries, root_subject_types, rules)
+
     with self.assertRaises(ValueError) as cm:
-      validator.validate()
+      scheduler.assert_ruleset_valid()
 
     self.assert_equal_with_printing(dedent("""
                      Rules with errors: 1
-                       (A, (Select(B),), noop):
-                         no matches for Select(B) with subject types: SubA
+                       (Exactly(A), (Select(Exactly(B)),), noop):
+                         no matches for Select(Exactly(B)) with subject types: SubA
                      """).strip(),
                                     str(cm.exception))
 
   def test_ruleset_with_rule_with_two_missing_selects(self):
     rules = [(A, (Select(B), Select(C)), noop)]
-    validator = RulesetValidator(RuleIndex.create(rules, tuple()),
-      goal_to_product={},
-      root_subject_types={SubA})
+    validator = self.create_validator({}, tuple(), {SubA}, rules)
     with self.assertRaises(ValueError) as cm:
       validator.validate()
 
     self.assert_equal_with_printing(dedent("""
                      Rules with errors: 1
-                       (A, (Select(B), Select(C)), noop):
-                         no matches for Select(B) with subject types: SubA
-                         no matches for Select(C) with subject types: SubA
+                       (Exactly(A), (Select(Exactly(B)), Select(Exactly(C))), noop):
+                         no matches for Select(Exactly(B)) with subject types: SubA
+                         no matches for Select(Exactly(C)) with subject types: SubA
                      """).strip(),
       str(cm.exception))
 
   def test_ruleset_with_selector_only_provided_as_root_subject(self):
     rules = [(A, (Select(B),), noop)]
-    validator = RulesetValidator(RuleIndex.create(rules, tuple()),
-      goal_to_product={},
-      root_subject_types={B,})
+    validator = self.create_validator({}, tuple(), {B, }, rules)
```

Comment:
here and elsewhere in this file some of your set literals contain trailing `,`'s and don't in other cases. afaik, these are never needed (e.g. not like `tuple`'s single-item-tuple-requires-trailing-comma). reads a bit awkwardly.